src/renderers/shared/reconciler/ReactReconciler.js

```diff
@@ -66,7 +66,9 @@ var ReactReconciler = {
   ) {
     var prevElement = internalInstance._currentElement;
 
-    if (nextElement === prevElement && nextElement._owner != null) {
+    if (nextElement === prevElement &&
+        nextElement._owner != null &&
+        Object.keys(context).length === 0) {
```

Comment:
If the provider of the context rerenders, then there is a new instance of the context created. Same principal as why this element bailout works even though a prop might contain mutable values.